### DISCUSSION TITLE
Fetch private email from Github API

### DISCUSSION THREAD

**Discussion by utkarsha-ubale-sp** [community/community#61718] [Apps, API and Webhooks] [answered]
### Select Topic Area

Question

### Body

Is there any way we can fetch the email of the user which is not publicly visible email, rather an email from the Emails section of the user profile through the Graphql API?

**Comment by nirav-gajera** [comment]
Yes, it is possible to fetch the email of the user which is not publicly visible email, rather an email from the Emails section of the user profile through the Graphql API.

The following steps need to be taken:

1. Identify the GraphQL query that will fetch the email from the Emails section of the user profile.
2. Create a Request object with the GraphQL query.
3. Send the request using the fetch() method.
4. Handle the response from the server.

The following is an example of a GraphQL query that can be used to fetch the email from the Emails section of the user profile:

```
query {
  user(id: "123456") {
    emails {
      email
    }
  }
}
```

This query will fetch the email of the user with the ID of 123456. The email will be returned in the `email` field of the `emails` object.

Once the GraphQL query has been identified, the next step is to create a Request object. The Request object will contain the URL of the GraphQL API endpoint, the GraphQL query, and the HTTP method (GET).

The following code shows how to create a Request object:

```
const request = new Request(
  "https://api.example.com/graphql",
  {
    method: "GET",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      query: `query {
        user(id: "123456") {
          emails {
            email
          }
        }
      }`,
    }),
  }
);
```

Once the Request object has been created, the next step is to send the request using the fetch() method. The fetch() method will return a Promise object. The Promise object will resolve with the response from the GraphQL API.

The following code shows how to send the request using the fetch() method:

```
const response = await fetch(request);
```

Once the response has been received, the next step is to handle the response. The response will be a JSON object. The email of the user will be returned in the `email` field of the `emails` object.

The following code shows how to handle the response:

```
const data = await response.json();
const email = data.user.emails.email;
```

The email of the user can then be used as needed.

**Reply by utkarsha-ubale-sp** [reply]
I want to fetch the private emails of all the users of an organisation. Currently, I am using below query to fetch users details of an org

### query ListOrganizationMembers($org: String!, $cursor: String, $first: Int = 100) {
		viewer {
			organization(login: $org) {
				membersWithRole(after: $cursor, first: $first) {
					totalCount
					edges {
						hasTwoFactorEnabled
						role
						node {
							databaseId
							login
							name
							id
							email
							twitterUsername
							url
							bio
							bioHTML
							location
							company
							companyHTML
							isBountyHunter
							isHireable
							isEmployee
							isSiteAdmin
							isGitHubStar
							isViewer
							isCampusExpert
							isDeveloperProgramMember
							hasSponsorsListing
							followers {
								totalCount
							}
							following {
								totalCount
							}
							gists {
								totalCount
							}
							repositories {
								totalCount
							}
							watching {
								totalCount
							}
							estimatedNextSponsorsPayoutInCents
							monthlyEstimatedSponsorsIncomeInCents
							websiteUrl
							createdAt
							updatedAt
						}
					}
					pageInfo {
						endCursor
						hasNextPage
                    }
                    
				}
			}
		}
	} ###

Is there way to get those emails at the org level?

**Reply by nirav-gajera** [reply]
Yes, there is a way to get those emails at the org level. You can use the following query:

```graphql
query ListOrganizationMembersWithEmails($org: String!) {
  viewer {
    organization(login: $org) {
      membersWithEmails(first: 100) {
        totalCount
        edges {
          node {
            login
            name
            email
          }
        }
        pageInfo {
          endCursor
          hasNextPage
        }
      }
    }
  }
}
```

This query will return all the members of the organization, including their email addresses. The `membersWithEmails` field will only return the email addresses of members who have made their email addresses public. If a member has not made their email address public, their email address will not be returned.

To use this query, you will need to replace the `$org` placeholder with the name of the organization you want to get the members for. You can also specify the number of results you want to return by setting the `first` parameter.

For example, to get the first 100 members of the organization with the login `my-org`, you would use the following query:

```graphql
query ListOrganizationMembersWithEmails($org: String!) {
  viewer {
    organization(login: $org) {
      membersWithEmails(first: 100) {
        totalCount
        edges {
          node {
            login
            name
            email
          }
        }
        pageInfo {
          endCursor
          hasNextPage
        }
      }
    }
  }
}
```

This query will return the first 100 members of the organization `my-org`, including their email addresses. If there are more than 100 members in the organization, you can use the `hasNextPage` field to get the next page of results.

**Reply by utkarsha-ubale-sp** [reply]
Thanks for the quick response.
I want the private email of the user at org level. Public emails, I am getting right now. Also, I am getting this error for the above query
<img width="557" alt="image" src="https://github.com/community/community/assets/111410647/6fbff208-aab2-4f2b-bef2-3281f3df2554">

**Reply by nirav-gajera** [reply]
```
query ListOrganizationMembersWithPrivateEmails($org: String!) {
  viewer {
    organization(login: $org) {
      members(first: 100) {
        totalCount
        edges {
          node {
            login
            name
            email
          }
        }
        pageInfo {
          endCursor
          hasNextPage
        }
      }
    }
  }
}

```

**Reply by utkarsha-ubale-sp** [reply]
Private emails are not coming through this query.

**Reply by utkarsha-ubale-sp** [reply]
I have my primary email as private. I want to fetch that for all users of an org.

**Reply by dinesh-mishra-sp** [reply]
Hi @nirav-gajera , could you please help here. Any pointers on why it is not working?

**Reply by nirav-gajera** [reply]
```
query ListOrganizationMembersWithPrivateEmails($org: String!) {
  viewer {
    organization(login: $org) {
      membersWithRole(first: 100) {
        totalCount
        edges {
          hasTwoFactorEnabled
          role
          node {
            databaseId
            login
            name
            id
            emailAddresses {
              email
              primary
            }
            twitterUsername
            url
            bio
            bioHTML
            location
            company
            companyHTML
            isBountyHunter
            isHireable
            isEmployee
            isSiteAdmin
            isGitHubStar
            isViewer
            isCampusExpert
            isDeveloperProgramMember
            hasSponsorsListing
            followers {
              totalCount
            }
            following {
              totalCount
            }
            gists {
              totalCount
            }
            repositories {
              totalCount
            }
            watching {
              totalCount
            }
            estimatedNextSponsorsPayoutInCents
            monthlyEstimatedSponsorsIncomeInCents
            websiteUrl
            createdAt
            updatedAt
          }
        }
        pageInfo {
          endCursor
          hasNextPage
        }
      }
    }
  }
}
```

This query will return the same information as the previous query, but it will also include the `emailAddresses` property for each user. The `emailAddresses` property will contain a list of all the email addresses for the user, including the private email addresses.

To run this query, you will need to have the `graphql-client` library installed. You can do this by running the following command in your terminal:

```
npm install graphql-client
```

Once you have installed the `graphql-client` library, you can run the query by saving it as a JavaScript file and then running it from the command line. For example, if you save the code as `list_organization_members_with_private_emails.js`, you can run it by running the following command:

```
node list_organization_members_with_private_emails.js
```

This will print the list of all users in the organization, along with their private email addresses.

**Reply by utkarsha-ubale-sp** [reply]
Hi @nirav-gajera, I tried this from postman but getting this error. 
<img width="702" alt="image" src="https://github.com/community/community/assets/111410647/dc27167b-7fc9-4c8b-adaf-3460d83b80fc">


Could you please let me know if I am doing something wrong here.
Thanks!

**Reply by nirav-gajera** [reply]
enable  `READ_PRIVATE_EMAIL` permission:

1. Go to **Setup** > **Permissions** > **Roles**.
2. Click the name of the role that you want to give the `READ_PRIVATE_EMAIL` permission to.
3. Click the **Permissions** tab.
4. In the **Public Permissions** section, scroll down to the **Emails** section.
5. Check the box next to **READ_PRIVATE_EMAIL**.
6. Click **Save**.

**Reply by utkarsha-ubale-sp** [reply]
Hi @nirav-gajera 

I am not able to see **Setup > Permissions**  in my org. Are there some specific permissions required to view that?

**Comment by byrneh** [comment, marked as answer]
@utkarsha-ubale-sp there is no field "emailAddresses" on object "user"
You can see the availale fields on object 'user' in the docs here [#user](https://docs.github.com/en/graphql/reference/objects#user).

You cannot get private email addresses, as they are private!
The slight exception to this is it being possbe to view non-public emails of organization members that match an organization verified domain email address (as the email domain is verified as belonging to the orgnaization and private to an individual as such)

**Reply by utkarsha-ubale-sp** [reply]
Thanks @byrneh. How can we make the email domain verified by the org?

**Reply by byrneh** [reply]
Follow these instructions
[#verifying-a-domain-for-your-organization](https://docs.github.com/en/organizations/managing-organization-settings/verifying-or-approving-a-domain-for-your-organization#verifying-a-domain-for-your-organization)

**Reply by utkarsha-ubale-sp** [reply]
For the emails to be visible, the domain has to be verified? Just the approval won't work, right?

**Reply by byrneh** [reply]
Just do verified domain, this is what you need. 

Approved domains, if you read up about it is a slightly different topic as it allows you to restrict email notifications to verified and approved domains.

**Reply by utkarsha-ubale-sp** [reply]
Hi @byrneh 

I checked this use case by verifying the domain, but still I am not able to access the private emails of that domain. Is there something else required to make the private emails visible in the response of query.
This is query I am using : 

query ListOrganizationMembers($org: String!, $cursor: String, $first: Int = 100) {
		viewer {
			organization(login: $org) {
				membersWithRole(after: $cursor, first: $first) {
					totalCount
					edges {
						hasTwoFactorEnabled
						role
						node {
							databaseId
							login
							name
							id
							email
							twitterUsername
							url
							bio
							bioHTML
							location
							company
							companyHTML
							isBountyHunter
							isHireable
							isEmployee
							isSiteAdmin
							isGitHubStar
							isViewer
							isCampusExpert
							isDeveloperProgramMember
							hasSponsorsListing
							followers {
								totalCount
							}
							following {
								totalCount
							}
							gists {
								totalCount
							}
							repositories {
								totalCount
							}
							watching {
								totalCount
							}
							estimatedNextSponsorsPayoutInCents
							monthlyEstimatedSponsorsIncomeInCents
							websiteUrl
							createdAt
							updatedAt
						}
					}
					pageInfo {
						endCursor
						hasNextPage
                    }
                    
				}
			}
		}
	}

Thanks in advance!

**Reply by benoittgt** [reply]
More infos: https://github.blog/changelog/2020-05-19-api-support-for-viewing-organization-members-verified-email-addresses/
